src/action-menu/_ActionUtils.jsx

```diff
@@ -0,0 +1,193 @@
+      onArrow(key, 'first');
+    } else if (event.nativeEvent.keyCode === KEY_END) {
+      shouldPrevent = true;
+      onArrow(key, 'last');
+    } else {
+      const char = event.key;
+      if (isValidChar(char)) {
+        shouldPrevent = true;
+        onChar(key, char);
+      }
+    }
+
+    if (shouldPrevent) {
+      event.preventDefault();
+      event.stopPropagation();
+    }
+  }
+);
+
+/**
+ * Returns the items with their first character matching the char parameter.
+ * @param {string} char The char to search upon.
+ * @param {array} items The flattened array of child items.
+ */
+const itemsByChar = (char, items) => items.filter(item => item.label.charAt(0).toLowerCase() === char.toLowerCase());
+
+/**
+ * Disables focus styles for the target of the given event. Typically used as an onMouseDown callback on selectable elements.
+ * @param {string} key The action key to search by.
+ * @param {array} items The flattened array of child items.
+ */
+const indexOfKey = (key, items) => items.findIndex(item => item.actionKey === key);
+
+/**
+ * Returns a flattened array of enabled items in the order they are visible.
+ * @param {string} key The currently active item key.
+ * @param {array} items The flattened array of child items.
+ * @param {string} direction The child node to strip the items from.
+ */
+const itemByDirection = (key, items, direction) => {
+  const currentIndex = indexOfKey(key, items);
+  const maxIndex = items.length - 1;
+
+  let newIndex;
+  if (direction === 'first') {
+    newIndex = 0;
+  } else if (direction === 'last') {
+    newIndex = maxIndex;
+  } else if (direction === 'next') {
+    if (currentIndex === maxIndex || currentIndex < 0) {
+      newIndex = 0;
+    } else {
+      newIndex = currentIndex + 1;
+    }
+  } else if (direction === 'previous') {
+    if (currentIndex <= 0) {
+      newIndex = maxIndex;
+    } else {
+      newIndex = currentIndex - 1;
+    }
+  }
+
+  return items[newIndex];
+};
+
+/**
+ * Returns the next matching item for the char and currently active key.
+ * @param {string} key The currently active item key.
+ * @param {array} items The flattened array of child items.
+ * @param {string} char The current char from keystroke.
+ */
+const itemByChar = (key, items, char) => {
+  const charMatches = itemsByChar(char, items);
+  const matchCount = charMatches.length;
+  if (!matchCount) {
+    return undefined;
+  }
+
+  let newIndex = 0;
+  const currentIndex = indexOfKey(key, charMatches);
+  if (matchCount > 1 && currentIndex >= 0 && currentIndex < matchCount - 1) {
+    newIndex = currentIndex + 1;
+  }
+
+  return charMatches[newIndex];
+};
+
+/**
+ * Returns a flattened array of enabled items in the order they are visible.
+ * @param {node} children The child node to strip the items from.
+ */
+const flattenActionItems = (children, showSelection = false) => {
+  let tempSelection = showSelection;
+  const actionItems = [];
+  React.Children.forEach(children, child => {
```

Comment:
we could use `map` and slightly clean this up by directly returning an array instead of building it via pushes
```
const items = React.Children.map(children, (child) => { }
```

Reply:
I think the issue with using map would be that the result would be peppered with undefined values resulting from that presence checks here. We could certainly filter it down afterwards, but that would require two iterations over the array vs the one here.